src/Core/ExecutionPathTracer/ExecutionPathTracer.cs

```diff
@@ -0,0 +1,307 @@
+        /// Retrieves the <c>QubitRegister</c> associated with the given <c>Qubit</c> or create a new
+        /// one if it doesn't exist.
+        /// </summary>
+        private QubitRegister GetQubitRegister(Qubit qubit)
+        {
+            if (!this.qubitRegisters.ContainsKey(qubit.Id))
+            {
+                this.qubitRegisters[qubit.Id] = new QubitRegister(qubit.Id);
+            }
+
+            return this.qubitRegisters[qubit.Id];
+        }
+
+        private List<QubitRegister> GetQubitRegisters(IEnumerable<Qubit> qubits)
+        {
+            return qubits.Select(this.GetQubitRegister).ToList();
+        }
+
+        /// <summary>
+        /// Creates a new <c>ClassicalRegister</c> and associate it with the given <c>Qubit</c>.
+        /// </summary>
+        private ClassicalRegister CreateClassicalRegister(Qubit measureQubit)
+        {
+            var qId = measureQubit.Id;
+
+            if (!this.classicalRegisters.ContainsKey(qId))
+            {
+                this.classicalRegisters[qId] = new List<ClassicalRegister>();
+            }
+
+            var cId = this.classicalRegisters[qId].Count;
+            ClassicalRegister register = new ClassicalRegister(qId, cId);
+
+            // Add classical register under the given qubit id
+            this.classicalRegisters[qId].Add(register);
+
+            return register;
+        }
+
+        /// <summary>
+        /// Retrieves the most recent <c>ClassicalRegister</c> associated with the given <c>Qubit</c>.
+        /// </summary>
+        /// <remarks>
+        /// Currently not used as this is intended for classically-controlled operations.
+        /// </remarks>
+        private ClassicalRegister GetClassicalRegister(Qubit controlQubit)
+        {
+            var qId = controlQubit.Id;
+            if (!this.classicalRegisters.ContainsKey(qId) || this.classicalRegisters[qId].Count == 0)
+            {
+                throw new Exception("No classical registers found for qubit {qId}.");
+            }
+
+            // Get most recent measurement on given control qubit
+            var cId = this.classicalRegisters[qId].Count - 1;
+            return this.classicalRegisters[qId][cId];
+        }
+
+        /// <summary>
+        /// Given a <c>Type</c> and its value, extract its fields and format it as a string.
+        /// </summary>
+        private string? ExtractArgs(Type t, object value)
+        {
+            List<string?> fields = new List<string?>();
+
+            foreach (var f in t.GetFields())
+            {
+                // If field is a tuple, recursively extract its inner arguments and format as a tuple string.
+                if (f.FieldType.IsTuple())
+                {
+                    var nestedArgs = f.GetValue(value);
+                    if (nestedArgs != null)
+                    {
+                        var nestedFields = this.ExtractArgs(f.FieldType, nestedArgs);
+                        fields.Add(nestedFields);
```

Comment:
Having a `.Add` method in a foreach loop is a great code smell for an opportunity to use `Select` instead. The trick here will be that you may want SelectMany to flatten out tuples...

Reply:
Ah, thanks for the tip! Still getting used to the tricks of C# :)